### ISSUE TITLE
`misc-const-correctness` does not warn when using `auto` to declare variable

### ISSUE THREAD

**Issue by ChrisThrasher** [llvm/llvm-project#60789]
```cpp
void foo() {
    bool a = true; // recommends adding const
    auto b = true; // no diagnostic emitted
}
```

When declaring a local variable with `const`, `misc-const-correctness` will never recommend adding `const`. I've tried for built-in types as well as user-defined types and can't under any circumstances get it to recommend adding `const` to an `auto` type.

**Comment by ChrisThrasher** [comment]
https://github.com/llvm/llvm-project/blob/8a386b22509d4ad1fccef3cf8fb42c887ec29874/clang-tools-extra/clang-tidy/misc/ConstCorrectnessCheck.cpp#L88

Maybe this line is to blame? I believe it's saying to not match local variables whose initializer is instantiation dependent? `auto` local vars may have different types based on the expression initializing them and perhaps this causes issues with this check? That would imply that this behavior is intention and not incidental.

**Comment by ChrisThrasher** [comment]
I hacked on this check to figure out what's to blame. The issue is the presence of `hasType(autoType())` which means the author of this check explicitly chose to ignore flagging `auto` types. I worry this may be an intentional choice but I'm still wondering why this check may not be appropriate in such circumstances.

https://github.com/llvm/llvm-project/blob/8a386b22509d4ad1fccef3cf8fb42c887ec29874/clang-tools-extra/clang-tidy/misc/ConstCorrectnessCheck.cpp#L77

**Comment by ChrisThrasher** [comment]
https://reviews.llvm.org/D54943

I found the review for this check and could not find any discussion about why `auto` types were omitted from analysis. I used my patched version of clang-tidy to run this check against SFML to confirm that it does in fact work. It found many more places to add `const`, none of which seem incorrect.

https://github.com/SFML/SFML/pull/2415/commits/6f8b4eb0fd53ad19d86933168e433349e667312c

@JonasToth Do you have any insight on this?

**Comment by ChrisThrasher** [comment]
https://github.com/ChrisThrasher/llvm-project/commit/acefd712142b69d61b1d1755e015f2a392c78f16

Here's my patch.

**Comment by EugeneZelenko** [comment]
@ChrisThrasher: See https://llvm.org/docs/Phabricator.html for patch process.

**Comment by ChrisThrasher** [comment]
I'm doing my due diligence before submitting a patch in case this was an intentional decision.

**Comment by ChrisThrasher** [comment]
https://reviews.llvm.org/D145856

**Comment by ambyld** [comment]
Did this fix get merged?

**Comment by ChrisThrasher** [comment]
No
